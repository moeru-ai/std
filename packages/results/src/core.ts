export interface Err<E> {
  __type__: 'err'
  error: E
}

export interface None {
  __type__: 'none'
}

export interface Ok<T> {
  __type__: 'ok'
  value: T
}

export type Option<T> = None | Some<T>

export type Result<T, E> = Err<E> | Ok<T>

export interface Some<T> {
  __type__: 'some'
  value: T
}

export const err = <T, E>(error: E): Result<T, E> => ({
  __type__: 'err',
  error,
})

export const none: Option<never> = {
  __type__: 'none',
}

export const ok = <T, E>(value: T): Result<T, E> => ({
  __type__: 'ok',
  value,
})

export const some = <T>(value: T): Option<T> => ({
  __type__: 'some',
  value,
})
