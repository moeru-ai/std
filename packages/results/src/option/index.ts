export * from './and'
export * from '../core'
export * from './extract'
export * from './is'
export * from './map'
export * from './match'
export * from './ok'
export * from './or'
export * from './unwrap'
export * from './wrap'
