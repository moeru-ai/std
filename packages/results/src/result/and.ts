import type { Ok, Result } from '../core'

import { isOk } from './is'

export const and = <T, E>(r: Result<T, E>, fallback: Result<T, E>): Result<T, E> =>
  isOk(r)
    ? r
    : fallback

export const andThen = <T1, T2, E>(r: Result<T1, E>, onOk: (o: Ok<T1>) => Result<T2, E>): Result<T2, E> =>
  isOk(r)
    ? onOk(r)
    : r

// eslint-disable-next-line sonarjs/no-identical-functions
export const andThenAsync = async <T1, T2, E>(r: Result<T1, E>, onOk: (o: Ok<T1>) => Promise<Result<T2, E>>): Promise<Result<T2, E>> =>
  isOk(r)
    ? onOk(r)
    : r
