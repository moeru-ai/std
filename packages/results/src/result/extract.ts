import type { Result } from '../core'

import { isOk } from './is'

/** @experimental */
export const extract = <T, E>(r: Result<T, E>): E | T =>
  isOk(r)
    ? r.value
    : r.error
