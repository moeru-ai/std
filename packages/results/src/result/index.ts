export * from './and'
export * from './expect'
export * from './extract'
export * from './from'
export * from './is'
export * from './map'
export * from './match'
export * from './option'
export * from './or'
export * from './unwrap'
export * from './wrap'
