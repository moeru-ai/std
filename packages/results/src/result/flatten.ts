import type { Result } from '../core'

import { err } from '../core'
import { match } from './match'

export const flatten = <T, E>(r: Result<Result<T, E>, E>): Result<T, E> => match(
  r,
  inner => inner,
  e => err(e),
)
