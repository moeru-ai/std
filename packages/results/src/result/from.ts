import type { Result } from '../core'

import { err, ok } from '../core'

/** @experimental */
export const from = <T, E = Error>(cb: () => T): Result<T, E> => {
  try {
    return ok(cb())
  }
  catch (e) {
    // eslint-disable-next-line @masknet/type-prefer-return-type-annotation
    return err(e) as Result<T, E>
  }
}
