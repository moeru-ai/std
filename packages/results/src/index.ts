import type { None, Option, Some } from './option'
import type { Err, Ok, Result } from './result'

import * as option from './option'
import * as result from './result'

export {
  Err,
  None,
  Ok,
  Option,
  option,
  result,
  Result,
  Some,
}
