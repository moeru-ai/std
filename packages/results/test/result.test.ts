import { describe, expect, it } from 'vitest'

import type { Result } from '../src'

import { err, match, ok, unwrap, wrap } from '../src/result'

describe('@moeru/results/result', () => {
  enum MathError {
    DivisionByZero = 'DivisionByZero',
    NegativeSquareRoot = 'NegativeSquareRoot',
    NonPositiveLogarithm = 'NonPositiveLogarithm',
  }

  type MathResult = Result<number, MathError>

  let div = (x: number, y: number): MathResult =>
    y === 0
      ? err(MathError.DivisionByZero)
      : ok(x / y)

  let sqrt = (x: number): MathResult =>
    x < 0
      ? err(MathError.NegativeSquareRoot)
      : ok(Math.sqrt(x))

  let ln = (x: number): MathResult =>
    x <= 0
      ? err(MathError.NonPositiveLogarithm)
      : ok(Math.log(x))

  // https://doc.rust-lang.org/rust-by-example/std/result.html
  it('basic', () => {
    let op = (x: number, y: number): number =>
      match(
        div(x, y),
        radio => match(
          ln(radio),
          // eslint-disable-next-line sonarjs/no-nested-functions
          ln => match(
            sqrt(ln),
            sqrt => sqrt,
            (err) => { throw err },
          ),
          // eslint-disable-next-line sonarjs/no-nested-functions
          (err) => { throw err },
        ),
        (err) => { throw err },
      )

    expect(() => op(1, 10)).toThrowErrorMatchingSnapshot()
    expect(() => op(1, 0)).toThrowErrorMatchingSnapshot()
  })

  // https://doc.rust-lang.org/rust-by-example/std/result/question_mark.html
  it('wrap', () => {
    let op = (x: number, y: number): number => match(
      wrap(() => {
        let ratio = unwrap(div(x, y))
        let _ln = unwrap(ln(ratio))
        return sqrt(_ln)
      }),
      value => value,
      (err) => { throw err },
    )

    expect(() => op(1, 10)).toThrowErrorMatchingSnapshot()
    expect(() => op(1, 0)).toThrowErrorMatchingSnapshot()
  })
})
