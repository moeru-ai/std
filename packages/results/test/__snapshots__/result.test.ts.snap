// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`@moeru/results/result > basic 1`] = `"NegativeSquareRoot"`;

exports[`@moeru/results/result > basic 2`] = `"DivisionByZero"`;

exports[`@moeru/results/result > wrap 1`] = `"NegativeSquareRoot"`;

exports[`@moeru/results/result > wrap 2`] = `[Error: called \`result.unwrap()\` on an \`Err\` value]`;
